submission/races.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package races
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/ctpolicy"
+)
+
+const (
+	// PostBatchInterval is duration between parallel batch call and subsequent requests to Logs within group.
+	// TODO(Mercurrent): optimize to avoid excessive requests.
+	PostBatchInterval = time.Second
+)
+
+// submissionResult holds outcome of a single-log submission.
+type submissionResult struct {
+	sct *ct.SignedCertificateTimestamp
+	err error
+}
+
+type groupState struct {
+	Name    string
+	Success bool
+}
+
+// safeSubmissionState is a submission state-machine for set of Log-groups. When some group is complete cancels all requests that are not needed by any group.
+type safeSubmissionState struct {
+	logToGroups map[string]map[string]bool
+	groupNeeds  map[string]int
+
+	results map[string]*submissionResult
+	cancels map[string]context.CancelFunc
+	mu      sync.Mutex
+}
+
+func newSafeSubmissionState(groups ctpolicy.LogPolicyData) *safeSubmissionState {
+	var s safeSubmissionState
+	s.logToGroups = ctpolicy.GroupByLogs(groups)
+	s.groupNeeds = make(map[string]int)
+	for _, g := range groups {
+		s.groupNeeds[g.Name] = g.MinInclusions
+	}
+	s.results = make(map[string]*submissionResult)
+	s.cancels = make(map[string]context.CancelFunc)
+	return &s
+}
+
+// request includes empty submissionResult in the set, returns whether the entry wasn't requested (added) before.
+func (sub *safeSubmissionState) request(logURL string, cancel context.CancelFunc) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sub.results[logURL] != nil {
```

Comment:
Explicit presence test might be clearer `if _, exists = sub.results[logURL]; exists { ... }`

Reply:
Done.